src/pytorch_lightning/strategies/ddp.py

```diff
@@ -296,39 +305,45 @@ def optimizer_step(
     def configure_ddp(self) -> None:
         log.detail(f"{self.__class__.__name__}: configuring DistributedDataParallel")
         self.pre_configure_ddp()
+        assert isinstance(self.model, (pl.LightningModule, _LightningPrecisionModuleWrapperBase))
         self.model = self._setup_model(LightningDistributedModule(self.model))
         self._register_ddp_hooks()
 
-    def determine_ddp_device_ids(self):
+    def determine_ddp_device_ids(self) -> Optional[List[int]]:
         if self.root_device.type == "cpu":
             return None
         return [self.root_device.index]
 
-    def barrier(self, *args, **kwargs) -> None:
+    def barrier(self, *args: Any, **kwargs: Any) -> None:
         if not distributed_available():
             return
         if torch.distributed.get_backend() == "nccl":
             torch.distributed.barrier(device_ids=self.determine_ddp_device_ids())
         else:
             torch.distributed.barrier()
 
-    def broadcast(self, obj: object, src: int = 0) -> object:
+    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
         obj = [obj]
         if self.global_rank != src:
-            obj = [None]
+            obj = [None]  # type: ignore[list-item]
         torch.distributed.broadcast_object_list(obj, src, group=_group.WORLD)
         return obj[0]
 
     def pre_backward(self, closure_loss: Tensor) -> None:
         """Run before precision plugin executes backward."""
+        assert self.lightning_module is not None
         if not self.lightning_module.automatic_optimization:
+            assert isinstance(self.model, DistributedDataParallel)
```

Comment:
This line keeps failing on tests with `AssertionError`, it seems the self.model is `deepspeed.runtime.engine.DeepSpeedEngine` when the argument of trainer 'strategy' is 'deepspeed'. I'm not sure how to resolve it, but `prepare_for_backward` just returns when the self.model is not instance of `DistributedDataParallel` internally, so should i change like this?
```suggestion
        if isinstance(self.model, DistributedDataParallel) and not self.lightning_module.automatic_optimization:
```

Reply:
changed in [f9f330e](https://github.com/Lightning-AI/lightning/pull/13885/commits/f9f330e39f2ca0a4495e8e45486637eb9ea43790), [1e2533c](https://github.com/Lightning-AI/lightning/pull/13885/commits/1e2533c476c58015e5e2811421cdd17799191d49).